user/caribou/module/src/DSO9254AEventConverter.cc

```diff
@@ -0,0 +1,377 @@
+      /*
+      std::cout << "  dx " << dx.at(nch) << std::endl
+		<< "  x0 " << x0.at(nch) << std::endl
+		<< "  dy " << dy.at(nch) << std::endl
+		<< "  y0 " << y0.at(nch) << std::endl;
+      */
+
+      // need once per channel
+      std::vector<double> ped;
+      std::vector<double> amp;
+      std::vector<std::vector<double>> waves;
+      
+      
+      for( int s=0; s<sgmnt_count; s++){ // loop semgents
+
+     	// container for one waveform
+	std::vector<double> wave;
+	
+	// prepare histogram with corresponding binning
+	TH1D* hist = new TH1D(
+			      Form( "waveform_run%i_ev%i_ch%i_s%i", ev->GetRunN(), ev->GetEventN(),
+				    nch, s ),
+			      Form( "waveform_run%i_ev%i_ch%i_s%i", ev->GetRunN(), ev->GetEventN(),
+				    nch, s ),
+			      np.at(nch),
+			      x0.at(nch) - dx.at(nch)/2.,
+			      x0.at(nch) - dx.at(nch)/2. + dx.at(nch)*np.at(nch)
+			      );
+	hist->GetXaxis()->SetTitle("time [ns]");
+	hist->GetYaxis()->SetTitle("signal [V]");
+      
+	// read channel data
+	std::vector<int16_t> words;
+	words.resize(points_per_words); // rawdata contains 8 byte words, scope sends 2 byte words 
+	int16_t wfi = 0;
+	for(int i=block_posit + 3 + pream_words + (s+0)*chann_words/sgmnt_count;
+	        i<block_posit + 3 + pream_words + (s+1)*chann_words/sgmnt_count;
+	      //i<block_posit + 3 + pream_words+chann_words;
+	        i++){
+	  
+	  // coppy channel data from whole data block
+	  memcpy(&words.front(), &rawdata[i], points_per_words*sizeof(int16_t));
+	  
+	  for( auto word : words ){
+	    
+	    // fill vectors with time bins and waveform
+	    if( nch == 0 && s == 0 ){ // this we need only once
+	      time.push_back( wfi * dx.at(nch) + x0.at(nch) );
+	    }
+	    wave.push_back( (word * dy.at(nch) + y0.at(nch)) );
+	    
+	    
+	    // fill histogram
+	    hist->SetBinContent(  hist->FindBin( wfi * dx.at(nch) + x0.at(nch) ),
+	    			  (double)word * dy.at(nch) + y0.at(nch) );
+	    
+	    wfi++;
+	    
+	  } // words
+	  
+	} // waveform
+
+	
+	// store waveform vector and create vector entries for pedestal and amplitude
+	waves.push_back( wave );
+	ped.push_back( 0. );
+	amp.push_back( 0. );
+
+	// write and delete
+	if( generateRoot ){
+	  hist->Write();
+	}
+	delete hist;
+
+	
+      } // segments
+
+
+      // not only for each semgent but for each channel
+      wavess.push_back( waves );
+      peds.push_back( ped );
+      amps.push_back( amp );
+	
+
+      // update position for next iteration
+      block_posit += block_words+1;
+    } // for channel
+    
+  } // if numblock
+
+
+  // process waveform data
+  if( time.size() > 0 ){ // only if there are waveform data
+  
+
+    // re-iterate waveforms
+    // wavess.at(channel).at(segment).at(point)
+    // ped   .at(channel).at(segment)
+    for( int s = 0; s<peds.at(0).size(); s++ ){
+      for( int p = 0; p<time.size(); p++ ) {      	
+	
+	//try{ // hotfix in case off incomplete waveforms
```

Comment:
Is this still necessary? Do we see incomplete waveforms?

Reply:
i think not, i removed that